src/main/scala/fi/vm/sade/hakurekisteri/integration/cache/RedisUpdateConcurrencyHandler.scala

```diff
@@ -0,0 +1,100 @@
+                                           implicit val executionContext: ExecutionContext) {
+
+  private val logger = org.slf4j.LoggerFactory.getLogger(getClass)
+  private val waitingPromises: ConcurrentHashMap[K, List[Promise[Option[T]]]] = new ConcurrentHashMap[K, List[Promise[Option[T]]]]()
+
+  def initiateLoadingIfNotYetRunning(key: K,
+                                     loader: K => Future[Option[T]],
+                                     loadedValueStorer: (K, Future[T]) => Future[_],
+                                     keyPrefixer: K => String): Future[Option[T]] = {
+    val (newClientPromise, loadIsInProgressAlready) = storePromiseForThisRequest(key)
+    val prefixKey = keyPrefixer(key)
+    if (loadIsInProgressAlready) {
+      logger.debug(s"Load already in progress for key $prefixKey, waiting to be resolved.")
+    } else {
+      r.get(prefixKey).onComplete {
+        case Success(found@Some(_)) =>
+          logger.debug(s"Value with $prefixKey had appeared in cache, no need to retrieve it.")
+          resolvePromisesWaitingForValueFromCache(key, Success(found), removeWaiters = true)
+        case Success(None) =>
+          logger.debug(s"Starting retrieval of $prefixKey with loader function.")
+          retrieveNewvalueWithLoader(key, loader, loadedValueStorer)
+        case failure@Failure(e) =>
+          resolvePromisesWaitingForValueFromCache(key, failure, removeWaiters = true)
+      }
+    }
+    newClientPromise.future
+  }
+
+  private def storePromiseForThisRequest(key: K): (Promise[Option[T]], Boolean) = {
+    logger.debug("Adding new client promise:")
+    val updateFunction = new BiFunction[K, List[Promise[Option[T]]], List[Promise[Option[T]]]] {
+      override def apply(key: K, promises: List[Promise[Option[T]]]): List[Promise[Option[T]]] = {
+        val myPromise = Promise[Option[T]]
+        if (promises == null) {
+          List(myPromise)
+        } else {
+          myPromise :: promises
+        }
+      }
+    }
+    val (newClientPromise :: otherPromises): Seq[Promise[Option[T]]] = waitingPromises.compute(key, updateFunction)
+
+    val numberWaiting = otherPromises.size + 1
+    logger.debug(s"Waiting after adding: $numberWaiting")
+    if (numberWaiting > limitOfWaitingClientsToLog) {
+      logger.warn(s"Already $numberWaiting clients waiting for value of $key")
+    }
+    (newClientPromise, otherPromises.nonEmpty)
+  }
+
+  private def retrieveNewvalueWithLoader(key: K,
+                                         loader: K => Future[Option[T]],
+                                         loadedValueStorer: (K, Future[T]) => Future[_]): Unit = {
+    val loadingFuture = loader(key)
+    loadingFuture.onComplete { result =>
+      if (result.isSuccess) {
+        result.get match {
+          case Some(foundItem) =>
+            loadedValueStorer(key, Future.successful(foundItem)).onComplete { _ =>
+              resolvePromisesWaitingForValueFromCache(key, result, removeWaiters = true)
```

Comment:
`resolvePromisesWaitingForValueFromCache` kutsut joissa `removeWaiters = true` voisi ehkä kaikki vielä siirtää `initiateLoadingIfNotYetRunning` funktioon. Vähentäisi toistoa, ja paikkoja joissa implisiittinen lukko vapautetaan.

Reply:
Meinaatko jotenkin näin? https://github.com/Opetushallitus/hakurekisteri/commit/c4775cc775f4574ed268464c7d44393f517d0c8d